src/lib/accessibility/FocusZone/FocusTrapZone.tsx

```diff
@@ -0,0 +1,296 @@
+      focusAsync(_lastTabbableChild)
+      ev.preventDefault()
+      ev.stopPropagation()
+    } else if (!ev.shiftKey && _lastTabbableChild === ev.target) {
+      focusAsync(_firstTabbableChild)
+      ev.preventDefault()
+      ev.stopPropagation()
+    }
+  }
+
+  private _forceFocusInTrap = (ev: Event, triggeredElement: HTMLElement) => {
+    if (
+      FocusTrapZone._focusStack.length &&
+      this === FocusTrapZone._focusStack[FocusTrapZone._focusStack.length - 1]
+    ) {
+      if (!this._root.current.contains(triggeredElement)) {
+        this._findElementAndFocusAsync()
+        ev.preventDefault()
+        ev.stopPropagation()
+      }
+    }
+  }
+
+  private _handleOutsideFocus = (ev: FocusEvent): void => {
+    const focusedElement = document.activeElement as HTMLElement
+    focusedElement && this._forceFocusInTrap(ev, focusedElement)
+  }
+
+  private _handleOutsideClick = (ev: MouseEvent): void => {
+    const clickedElement = ev.target as HTMLElement
+    clickedElement && this._forceFocusInTrap(ev, clickedElement)
+  }
+
+  private _subscribeToEvents = () => {
+    const { forceFocusInsideTrap, isClickableOutsideFocusTrap } = this.props
+    if (forceFocusInsideTrap) {
+      eventStack.sub('focus', this._handleOutsideFocus, {
+        target: this.windowElement,
+        useCapture: true,
+      })
+    }
+
+    if (!isClickableOutsideFocusTrap) {
+      eventStack.sub('click', this._handleOutsideClick, {
+        target: this.windowElement,
+        useCapture: true,
+      })
+    }
+  }
+
+  private _unsubscribeFromEvents = () => {
+    const { forceFocusInsideTrap, isClickableOutsideFocusTrap } = this.props
+    if (forceFocusInsideTrap) {
+      eventStack.unsub('focus', this._handleOutsideFocus, {
+        target: this.windowElement,
+        useCapture: true,
+      })
+    }
+
+    if (!isClickableOutsideFocusTrap) {
+      eventStack.unsub('click', this._handleOutsideClick, {
+        target: this.windowElement,
+        useCapture: true,
+      })
+    }
+  }
+
+  private _getPreviouslyFocusedElementOutsideTrapZone = () => {
+    const { elementToFocusOnDismiss } = this.props
+    let previouslyFocusedElement = this._previouslyFocusedElementOutsideTrapZone
+
+    if (elementToFocusOnDismiss && previouslyFocusedElement !== elementToFocusOnDismiss) {
+      previouslyFocusedElement = elementToFocusOnDismiss
+    } else if (!previouslyFocusedElement) {
+      previouslyFocusedElement = document.activeElement as HTMLElement
+    }
+
+    return previouslyFocusedElement
+  }
+
+  private _hideContentFromAccessibilityTree = () => {
+    const bodyChildren = (document.body && document.body.children) || []
+
+    if (bodyChildren.length && !document.body.contains(this._root.current)) {
+      // In case popup render options will change
+      console.warn(
+        'Body does not contain trap zone element as expected. If it is done intentionally, please, make sure to update FocusTrapZone.',
```

Comment:
still, this message is a bit confusing - as we need just to ask client to use ensure that trap zone element is placed inside body. Lets express exactly the same thought in the warning message:

> Body does not contain trap zone element as expected - ensure that trapped DOM element is a child of <body />.